### PR TITLE
Use upstream cid

### PR DESCRIPTION
**Pull request by vmx**
Instead of a custom implementation of cid, wrap the upstream version.

**Summary of changes**
Changes introduced in this pull request:
- Using upstream rust-cid

**Other information and links**

I removed the "default CID" case from the CID serializer as the old default CID won't work with rust-cid as it was an invalid CID.

This PR is intentionally using the rust-cid master branch and not a release as I would like to see if things would work with that version of rust-cid or if any changes would be nice/needed before I do a release (as this release is a huge breaking one, so it's the chance to break more things).

There is a `new_from_cbor()` function which I made into a module's function instead of adding it to the `Cid` impl. The reason is that I wanted to keep the `Cid` a pure wrapper around the upstream `Cid` without additional methods. Though I'm happy to change that if requested.

Feedback in welcome.

### CODE REVIEW

File: ipld/cid/src/lib.rs [hunk 1]
```diff
@@ -35,28 +29,63 @@ const MULTIBASE_IDENTITY: u8 = 0;
...
 pub mod json;
 
-/// Representation of a IPLD CID.
-#[derive(PartialEq, Eq, Clone, Copy)]
-pub struct Cid {
-    pub version: Version,
-    pub codec: Codec,
-    pub hash: Multihash,
+pub const DAG_CBOR: u64 = 0x71;
+pub const DAG_PB: u64 = 0x70;
```
Comment (austinabell) [inline ipld/cid/src/lib.rs:33]: This doesn't have to be an exposed constant, it's not used for anything meaningful (just a test that could use any value)

File: ipld/src/lib.rs [hunk 2]
```diff
@@ -101,7 +101,7 @@ pub enum Ipld {
     /// let cid: Cid = "QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n".parse().unwrap();
     /// let v1 = ipld!(Link(cid.clone()));
     /// let v2 = ipld!(cid);
-    /// assert_eq!(v1, v2);
+    /// //assert_eq!(v1, v2);
```
Comment (austinabell) [inline ipld/src/lib.rs:104]: does this fail now?

**Review by austinabell — COMMENTED** [review]
Looks good! I'll just need to run interop tests and make sure no regressions as well as waiting for a few interop PRs to come in before merging this in though since this is a change that touches pretty much everything

File: ipld/cid/src/lib.rs (hunk 1 above)
Comment (vmx) [reply ipld/cid/src/lib.rs]: Changed tests to `DAG_CBOR`.

File: ipld/src/lib.rs (hunk 2 above)
Comment (vmx) [reply ipld/src/lib.rs]: That was an oversight during development. I doesn't fail.

**Comment by vmx** [comment]
@austinabell what about this plan:

 1. You run the integration tests. On success I'll cut a rust-cid release and update this PR. On failure I'll have a look into the failure (resp. helping you looking into).
 2. Once all PRs that should get merged before that one are in, you let me know and I'll then try to rebase this PR as quickly as possible, so that it doesn't block things.

**Comment by austinabell** [comment]
> @austinabell what about this plan:
> 
> 1. You run the integration tests. On success I'll cut a rust-cid release and update this PR. On failure I'll have a look into the failure (resp. helping you looking into).
> 2. Once all PRs that should get merged before that one are in, you let me know and I'll then try to rebase this PR as quickly as possible, so that it doesn't block things.

Sounds great to me, the PRs I would like to get in before retesting are #847 #849 and #851 and I will let you know when those come in. I do believe your changes should be good as is, as I did some testing of it, I just want to check a couple things first and be completely sure if you don't mind

File: ipld/cid/src/lib.rs [hunk 3]
```diff
@@ -105,121 +127,8 @@ impl<'de> de::Deserialize<'de> for Cid {
...
-            Version::V1 => self.to_string_v1(),
-        };
-        write!(f, "{}", encoded)
-    }
-}
-
-impl fmt::Debug for Cid {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "Cid(\"{}\")", self)
+        write!(f, "{}", self.0)
```
Comment (austinabell) [inline ipld/cid/src/lib.rs:132]: This print is showing up as a debug print and it's really hard to read in certain logs, can you change this back to printing the actual `to_string`?

File: ipld/blockstore/Cargo.toml [hunk 4]
```diff
@@ -6,6 +6,8 @@ edition = "2018"
 
 [dependencies]
 cid = { package = "forest_cid", path = "../cid" }
+#cid = { path = "/home/vmx/src/pl/rust/rust-cid", features = ["serde-codec"] }
+#cid = { path = "/home/vmx/src/pl/rust/rust-cid" }
```
Comment (austinabell) [inline ipld/blockstore/Cargo.toml:10]:
```suggestion
```

**Review by austinabell — COMMENTED** [review]
LGTM, should be good for release (these two comments are local changes)

**Review by austinabell — APPROVED** [review]

**Review by ec2 — APPROVED** [review]

**Comment by austinabell** [comment]
Thanks :)

**Pull request merged**